File: packages/mobile/test/direction_agreement_test.dart
// THE TWO PLATFORMS MUST NAME THE SAME LANGUAGES RTL.
//
// `directionOf` here and `RTL_PRIMARY` in `packages/gate/src/index.ts` answer
// the same question for the same product. They had drifted: both listed `ku`,
// which ICU calls LTR (bare `ku` is Kurmanji, Latin script — `ckb` is the
// Sorani that is RTL), and neither listed `prs`, which ICU calls RTL. The web
// additionally kept `az-arab`, `pa-arab` and `uz-arab` entries it could never
// reach, because it matched on the primary subtag alone.
//
// A comment saying "keep these in step" is not a mechanism. This reads the
// gate's list out of its TypeScript and fails when the two disagree — the same
// trick `latn_test.dart` uses to pin `data-lumo-latn` against `kLumoLatnIsland`,
// and the only other place the two platforms are checked against each other.
import 'dart:io';

import 'package:flutter/widgets.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:lumo_ui_mobile/lumo_ui_mobile.dart';

/// The gate's own set, read from source rather than copied.
Set<String> gateRtlPrimary() {
  final file = File('../gate/src/index.ts');
  expect(file.existsSync(), isTrue,
      reason: 'packages/gate/src/index.ts is where the web keeps the same list; '
          'if it moved, this test must follow it rather than be deleted');
  final source = file.readAsStringSync();
  final match =
      RegExp(r'const RTL_PRIMARY = new Set\(\[(.*?)\]\);', dotAll: true)
          .firstMatch(source);
  expect(match, isNotNull, reason: 'RTL_PRIMARY is no longer a Set literal');
  return RegExp('"([a-z-]+)"')
      .allMatches(match!.group(1)!)
      .map((m) => m.group(1)!)
      .toSet();
}

void main() {
  test('the mobile RTL set is the gate\'s RTL set', () {
    final gate = gateRtlPrimary();
    expect(gate, isNotEmpty);

    // Every tag the gate calls RTL, this platform must lay out RTL.
    for (final tag in gate) {
      expect(directionOf(tag), TextDirection.rtl, reason: 'gate says $tag is RTL');
    }

    // And nothing extra: a tag the gate omits must not be RTL here. Checked
    // through the public function, so the two tables cannot agree while the
    // LOOKUPS differ — which is how `az-Arab` was wrong on the web.
    for (final tag in const ['en', 'de', 'fr', 'tr', 'ku', 'az', 'pa', 'uz', 'ja', 'ru']) {
      expect(directionOf(tag), TextDirection.ltr, reason: '$tag is not in the gate set');
    }
  });

  test('Arabic script flips a language that is otherwise LTR', () {
    // The entries that were unreachable on the web for want of a script lookup.
    expect(directionOf('az'), TextDirection.ltr);
    expect(directionOf('az-Arab'), TextDirection.rtl);
    expect(directionOf('pa-Arab'), TextDirection.rtl);
    expect(directionOf('uz-Arab'), TextDirection.rtl);
    // Underscores are a Dart locale habit and must behave the same.
    expect(directionOf('uz_Arab'), TextDirection.rtl);
  });

  test('the two corrections ICU forced, named so they cannot quietly return', () {
    expect(directionOf('ku'), TextDirection.ltr, reason: 'Kurmanji is Latin script');
    expect(directionOf('ckb'), TextDirection.rtl, reason: 'Sorani is the RTL one');
    expect(directionOf('prs'), TextDirection.rtl, reason: 'Dari is Persian in Arabic script');
  });
}

File: packages/gate/src/direction-agreement.test.ts
/*
 * THE TWO ANSWERS TO ONE QUESTION MUST MATCH.
 *
 * `LumoHtml` writes `dir` from `core`'s `direction()`. `lang-dir` then reads
 * that attribute back and compares it against the gate's own profile. If the
 * two derive direction differently, the gate fails a page Lumo itself rendered
 * — and it did, for four tags:
 *
 *   ku        core ltr, gate rtl   (bare `ku` is Kurmanji, Latin script)
 *   az-Arab   core rtl, gate ltr   ┐ the table is keyed by PRIMARY SUBTAG, so
 *   pa-Arab   core rtl, gate ltr   │ its `az-arab`/`pa-arab`/`uz-arab` entries
 *   uz-Arab   core rtl, gate ltr   ┘ were unreachable
 *
 * Neither side was asked to change its mind: both now ask the platform, which
 * is what every other field of the gate's derived profile already did. This
 * file exists so the next divergence is a failing test rather than a consumer
 * discovering it.
 */
import { describe, expect, it } from "vitest";
import { direction } from "../../core/src/index.ts";
import { gradingFor, knownLocales } from "./index.ts";

/*
 * RTL languages, the Arabic-script variants of otherwise-LTR languages (the
 * class that was broken), and enough LTR tags that a rule of "always rtl" would
 * fail too.
 */
const TAGS = [
  "fa", "fa-IR", "ar", "ar-SA", "ar-EG", "he", "iw", "ur", "ps", "sd", "ug", "yi",
  "ckb", "dv", "nqo", "syr", "ks",
  "ku", "az", "az-Arab", "pa", "pa-Arab", "uz", "uz-Arab", "prs",
  "en", "en-US", "de", "de-CH", "fr", "ru", "zh", "zh-Hant-TW", "ja", "ko", "hi", "th", "tr", "pt-BR",
];

describe("core and the gate agree on direction", () => {
  it.each(TAGS)("%s", (tag) => {
    expect(gradingFor(tag).direction, `the gate and core disagree on ${tag}`).toBe(
      direction(tag as Parameters<typeof direction>[0]),
    );
  });

  it("including every locale the gate lists as built-in", () => {
    for (const tag of knownLocales()) {
      expect(gradingFor(tag).direction, `built-in ${tag}`).toBe(
        direction(tag as Parameters<typeof direction>[0]),
      );
    }
  });

  it("and the check itself can fail — rtl and ltr are both really produced", () => {
    // A test that only ever compares two constants proves nothing. Both
    // answers must appear in the set above, or this file is vacuous.
    const seen = new Set(TAGS.map((t) => gradingFor(t).direction));
    expect(seen).toEqual(new Set(["rtl", "ltr"]));
  });
});
